### ISSUE TITLE
irc relay not working with android app "Atomic"

### ISSUE THREAD

**Issue by github-cygwin** [weechat/weechat#1040]
Hi,

running weechat-1.9 on Fedora 25.  Yesterday I was trying for the first time to use
the irc relay, like this:

```
/relay add ipv4.irc.Freenode 9001
/set relay.network.password "pwd"
```
No ssl or anything for now since this is running over VPN only.

Then I configured the android IRC client "Atomic" (which I'm using with various
IRC servers for a good while) to connect to the weechat relay.  Connect succeeds,
but then it just hangs:

```
[12:05]  Connected to wee.free
[12:05]  Now logging in...
```
I activated relay raw output, and it turns out that weechat gets a request from
Atomic which it never replies to:

```
23.07 12:05 --> [1] irc.Freenode PASS pwd
23.07 12:05 --> [1] irc.Freenode CAP LS
23.07 12:05 NICK corinna
23.07 12:05 USER corinna 8 * :Corinna Vinschen
23.07 12:05 <-- [1] irc.Freenode :weechat.relay.irc CAP nick LS :server-time
23.07 12:05 --> [1] irc.Freenode CAP REQ :
```
Is there a way from the user side to make this work, or is that a problem in
weechat's relay which should reply to "CAP REQ :"?

Thanks,
Corinna

**Comment by github-cygwin** [comment]
From the source it seems that "CAP REQ :" will result in a call to
relay_irc_search_server_capability() with an empty string.  This
in turn returns -1 and weechat's relay will skip the CAP ACK
reply code, but not do anything else.

Shouldn't relay return CAP NAK in this scenario?

Thanks,
Corinna

**Comment by github-cygwin** [comment]
I just debugged this a little further and it seems that the real problem is that Atomic
doesn't send a "CAP END".  It seems to expect that the login continues after sending
"CAP REQ :".  I applied the following patch to weechat to workaround this:

```
diff --git a/src/plugins/relay/irc/relay-irc.c b/src/plugins/relay/irc/relay-irc.c
index 99fb0f3..f3253cf 100644
--- a/src/plugins/relay/irc/relay-irc.c
+++ b/src/plugins/relay/irc/relay-irc.c
@@ -1277,6 +1277,10 @@ relay_irc_recv_command_capab (struct t_relay_client *client,
                              (RELAY_IRC_DATA(client, nick)) ? RELAY_IRC_DATA(client, nick) : "nick",
                              str_capab);
         }
+       else if (!strcmp (argv[1], ":"))
+       {
+            RELAY_IRC_DATA(client, cap_end_received) = 1;
+       }
     }
     else if (weechat_strcasecmp (argv[0], "end") == 0)
     {
```
This lets Atomic continue.  That's a bit puzzeling to me, but apparently it works
with other servers like Freenode.

Corinna

**Comment by flashcode** [comment]
Hi,
Thanks for reporting the problem.
I fixed the parsing problem when there are no arguments on "CAP REQ", and now WeeChat answers with "ACK" only if all capabilities are OK, and "NAK" if at least one capability is not found (or in case of empty arguments).
Could you please try with latest version and ping me if there's still a problem with Atomic?
Thanks.

**Comment by github-cygwin** [comment]
Thanks for the patch, but this doesn't fix the problem.  It seems your patch and
my findings from my 2nd comment (Atomic expects that login proceeds after
sending "CAP REQ :") overlapped in terms of their timing.

Unfortunately I can't reopen the issue, only write this comment.  Hope that's
enough :)


Thanks,
Corinna

**Comment by flashcode** [comment]
For me it looks like there are 2 bugs in Atomic:

1. Sending `CAP REQ :` is weird and useless.
2. The client does not care about response and never sends `CAP END`.

I can fix on relay side, considering receiving `CAP REQ :` also means `CAP END` but this looks weird to me (it will work, but not the ideal solution for me).

**Comment by github-cygwin** [comment]
I agree that this is buggy.  I created a [github issue for Atomic](https://github.com/indrora/Atomic/issues/134) as well, and I'm not
sure what Atomic is doing there.

As far as the response goes, the log message from inside Atomic indicates that it
received the "CAP NAK".  However, this does not trigger a "CAP END" reply.

As I wrote, for some reason Atomic thinks sending the "CAP REQ :" is enough and
the login process should continue from the server side.  What strikes me as weird
here is that the missing "CAP END" does not have any negative impact on connections
to other IRC servers, apparently.  Directly connecting to Freenode or even the
Slack IRC gateway works as desired.  But then again, I really don't know the IRC
protocol all that well.

Thanks,
Corinna
